lib/elixir/lib/kernel.ex

```diff
@@ -957,7 +971,10 @@ defmodule Kernel do
       [1, 2 | 3]
 
   """
-  @spec (list ++ term) :: maybe_improper_list
+  @spec (nonempty_list ++ nonempty_list) :: nonempty_list
+  @spec (left ++ []) :: left when left: nonempty_list
+  @spec ([] ++ right) :: right when right: term
+  @spec (list ++ nonempty_improper_list(any, any)) :: nonempty_improper_list(any(), any())
```

Comment:
Is there the Erlang documentation for `:erlang.++/2`? All i was able to find was able to find is for `:lists.append/2`, 

> append(List1, List2) -> List3
> Types:
> List1 = List2 = List3 = [T]
> T = term()

Maybe I could submit a patch to Erlang.

The other thing is the spec is complecated, but it describes exactly what the function does.
Maybe when I include the documentation it will be easier to follow.